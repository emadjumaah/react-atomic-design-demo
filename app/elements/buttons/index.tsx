import React, { FunctionComponent } from 'react';
import { IDefaultProps } from 'types';
import styled from 'styled-components';

export enum ButtonVariant {
  PRIMARY = 'primary',
  SECONDARY = 'secondary',
  TERTIARY = 'tertiary',
}

export enum ButtonColor {
  BLUE = 'primary',
  ORANGE = 'secondary',
  GREEN = 'tertiary',
  WHITE = 'white',
  BLACK = 'black',
}

interface IButtonProps extends IDefaultProps {
  /**
   * [required] Children must be a string
   */
  children: string;
  /**
   * [required] Callback to trigger when user clicks on button
   */
  onClick: () => void;
  /**
   * [optional] Color of button
   *
   * @default ButtonColor.BLUE
   */
  color?: ButtonColor;
}

interface IBaseButtonProps extends IButtonProps {
  /**
   * [optional] Variant of button to use
   *
   * @default ButtonVariant.PRIMARY
   */
  variant?: ButtonVariant;
}

const BaseButton: FunctionComponent<IBaseButtonProps> = props => (
  <button data-testid="button" className={props.className} onClick={props.onClick}>
    {props.children}
  </button>
);

/**
 * @todo - Refactor how colors are used within application to support better accessibility
 * and possibly the ability to use tools like Polished's `lighten` or `darken` for better
 * hover & press effects with buttons
 */
const getBackgroundColor = (variant: ButtonVariant, color: ButtonColor) => {
  switch (variant) {
    case ButtonVariant.PRIMARY:
      return `
          background-color: var(--${color});
        `;
    case ButtonVariant.SECONDARY:
    case ButtonVariant.TERTIARY:
    default:
      return `
          background-color: transparent;
        `;
  }
};

const getBorderColor = (variant: ButtonVariant, color: ButtonColor) => {
  switch (variant) {
    case ButtonVariant.SECONDARY:
      return `
        border: 2px solid var(--${color});
      `;
    case ButtonVariant.PRIMARY:
    case ButtonVariant.TERTIARY:
    default:
      return `
        border: none
      `;
  }
};

const getFontColor = (variant: ButtonVariant, color: ButtonColor) => {
  switch (variant) {
    case ButtonVariant.PRIMARY:
      return color === ButtonColor.WHITE
        ? `color: var(--black);`
        : `color: var(--white);`;
    case ButtonVariant.SECONDARY:
    case ButtonVariant.TERTIARY:
    default:
      return `
        color: var(--${color});
      `;
  }
};

const getButtonCss = (props: IBaseButtonProps) => {
  const color = props.color || ButtonColor.BLUE;
  const variant = props.variant || ButtonVariant.PRIMARY;

  return `
    ${getBackgroundColor(variant, color)}
    ${getBorderColor(variant, color)}
    ${getFontColor(variant, color)}

    &:hover {
      cursor: pointer;
    }
  `;
};

const StyledBaseButton = styled(BaseButton)`
  ${props => getButtonCss(props)}
`;

export const PrimaryButton: FunctionComponent<IButtonProps> = props => (
  <StyledBaseButton {...props} variant={ButtonVariant.PRIMARY} />
);
export const SecondaryButton: FunctionComponent<IButtonProps> = props => (
  <StyledBaseButton {...props} variant={ButtonVariant.SECONDARY} />
);
export const TertiaryButton: FunctionComponent<IButtonProps> = props => (
  <StyledBaseButton {...props} variant={ButtonVariant.TERTIARY} />
);
