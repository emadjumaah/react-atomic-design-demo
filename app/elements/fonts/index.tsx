import React from 'react';
import { IDefaultProps } from 'types';

export const PageTitle = (props: IDefaultProps) => <h1 {...props} />;
export const SectionTitle = (props: IDefaultProps) => <h2 {...props} />;
export const SubsectionTitle = (props: IDefaultProps) => <h3 {...props} />;

interface IParagraphFontProps extends IDefaultProps {
  /**
   * [optional] Wraps children in `em` tag
   */
  emphasis?: boolean;
  /**
   * [optional] Wraps children in `strong` tag
   */
  strong?: boolean;
  /**
   * [optiona] Renders paragraph with Description instead of Primary styles
   */
  isDescription?: boolean;
}

const Emphasis = (props: any) => <em {...props} />;
const Strong = (props: any) => <strong {...props} />;
export const PrimaryText = (props: IParagraphFontProps) => {
  const className = props.isDescription
    ? `${props.className} description`
    : props.className;
  if (props.emphasis || props.strong) {
    const Em = props.emphasis ? Emphasis : React.Fragment;
    const St = props.strong ? Strong : React.Fragment;
    return (
      <p className={className}>
        <Em>
          <St>{props.children}</St>
        </Em>
      </p>
    );
  }
  return <p {...props} className={className} />;
};
export const DescriptionText = (props: IParagraphFontProps) => (
  <PrimaryText {...props} className={`${props.className} description`} />
);
