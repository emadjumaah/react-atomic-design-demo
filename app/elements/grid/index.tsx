import React, { FunctionComponent } from 'react';
import { IDefaultProps } from 'types';
import { Row } from 'layouts';

export type TRowColumn = [number, number];

export interface IGridProps extends IDefaultProps {
  /**
   * [required] Grid size [rows, columns]
   */
  grid: TRowColumn;
  /**
   * [required] Callback responsible for returning a valid JSX Element based
   * on current row/column position
   */
  renderCell: (row: number, column: number) => JSX.Element;
}

export const Grid: FunctionComponent<IGridProps> = props => {
  const rows = new Array(props.grid[0]).fill(0);
  const columns = new Array(props.grid[1]).fill(0);
  return (
    <>
      {rows.map((_, rIndex) => (
        <Row key={`row-${rIndex}`}>
          {columns.map((_, cIndex) => props.renderCell(rIndex, cIndex))}
        </Row>
      ))}
    </>
  );
};
