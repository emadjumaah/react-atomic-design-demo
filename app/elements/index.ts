export * from './fonts';
export * from './markdown';
export * from './navigation';
export * from './grid';
export * from './shapes';
export * from './buttons';
export * from './forms';
