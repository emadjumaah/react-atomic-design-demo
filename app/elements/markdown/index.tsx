import React, { FunctionComponent } from 'react';
import ReactMarkdown from 'react-markdown';
import { Column } from 'layouts';
import { markdownRenderers } from './components';

interface IMarkdownProps {
  /**
   * [required] Content source to convert from markdown to React components
   */
  source: string;
}

export const Markdown: FunctionComponent<IMarkdownProps> = props => (
  <Column>
    <ReactMarkdown {...props} renderers={markdownRenderers} />
  </Column>
);

export * from './components';
export * from './interactive-markdown';
