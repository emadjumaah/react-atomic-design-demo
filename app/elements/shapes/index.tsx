import React, { FunctionComponent } from 'react';
import styled from 'styled-components';
import { IDefaultProps } from 'types';
import { FlexSpace } from 'layouts';
import { Color } from 'config/colors';

interface IBaseShapeProps extends IDefaultProps {
  /**
   * [optional] Background color for shape
   *
   * @default Color.BLACK
   */
  backgroundColor?: Color;
}

const BaseShape: FunctionComponent<IBaseShapeProps> = props => <div {...props} />;

interface ICircleProps extends IBaseShapeProps {
  /**
   * [optional] Size of circle (diameter)
   *
   * @default FlexSpace.MEDIUM
   */
  size?: FlexSpace;
}

export const Circle = styled(BaseShape)<ICircleProps>`
  border-radius: 100%;
  height: ${props => props.size || FlexSpace.MEDIUM};
  width: ${props => props.size || FlexSpace.MEDIUM};
  margin: 8px;
  background-color: ${props => props.backgroundColor || Color.BLACK};
`;
