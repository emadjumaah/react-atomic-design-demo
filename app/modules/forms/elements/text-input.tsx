import React, { FunctionComponent, useContext, useEffect } from 'react';
import { ITextInputProps, TextInput as BaseTextInput } from 'elements';
import { FormContext, FormActions } from 'modules/forms';

interface IFormTextInput extends ITextInputProps {
  /**
   * [required] Unique identifier for text input within form
   */
  name: string;
}

export const TextInput: FunctionComponent<IFormTextInput> = ({ name, ...props }) => {
  const { dispatch } = useContext(FormContext);

  useEffect(() => {
    dispatch(FormActions.registerFormElement(name));
  }, []);

  return <BaseTextInput {...props} />;
};
