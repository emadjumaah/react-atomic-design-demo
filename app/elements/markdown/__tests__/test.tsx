import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Markdown, InteractiveMarkdown } from 'elements';

const allMarkdown = `
  # Heading 1
  ## Heading 1
  ### Heading 1
  # <a name="heading-with-anchor"></a>Heading With Anchor

  paragraph
  **Strong**
  _em_

  * Unordered list 1
  * Unordered list 2

  1. Ordered list 1
  2. Ordered list 2

  [Link to somethign](linktosomething.com)

  ---

  \`inline code\`
  \`\`\`
    const blockCodeFunc = () => {
      return 'works!';
    }
  \`\`\`
`;

const markdownContent = `Markdown Text`;
const onClick = jest.fn();
/**
 * @todo Replace with Button component
 */
const testId = 'button';
const clickableButton = (
  <button key={testId} data-testid={testId} onClick={onClick}>
    Click Me
  </button>
);

describe('Markdown', () => {
  it('Renders All Markdown Components', () => {
    const { container } = render(<Markdown source={allMarkdown} />);
    expect(container).toMatchSnapshot();
  });
  it('Renders Markdown with Interactive React Components', () => {
    const { container, getByTestId } = render(
      <InteractiveMarkdown>
        {[markdownContent, clickableButton]}
      </InteractiveMarkdown>,
    );
    expect(container).toMatchSnapshot();

    fireEvent.click(getByTestId(testId));
    expect(onClick).toHaveBeenCalledTimes(1);
  });
});
