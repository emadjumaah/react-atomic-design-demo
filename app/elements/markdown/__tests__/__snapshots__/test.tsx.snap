// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markdown Renders All Markdown Components 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 {
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border-radius: 0px;
}

.c6 {
  margin: 16px 0;
  background-color: var(--black);
  color: var(--white);
  border-radius: 3px;
  padding: 16px;
}

.c5 {
  color: var(--secondary);
}

.c4 {
  margin: 16px 0px;
  color: var(--black);
}

.c3 {
  margin-bottom: 16px;
  margin-top: 8px;
}

.c2 {
  margin-bottom: 16px;
  margin-top: 8px;
}

<div>
  <div
    class="c0 c1"
  >
    <h1>
      Heading 1
    </h1>
    <h2>
      Heading 1
    </h2>
    <h3>
      Heading 1
    </h3>
    <h1>
      <div>
        <a
          name="heading-with-anchor"
        />
      </div>
      <div />
      Heading With Anchor
    </h1>
    <p>
        paragraph
  
      <strong>
        Strong
      </strong>
      
  
      <em>
        em
      </em>
    </p>
    <ul
      class="c2"
    >
      <li>
        Unordered list 1
      </li>
      <li>
        Unordered list 2
      </li>
    </ul>
    <ol
      class="c3"
    >
      <li>
        Ordered list 1
      </li>
      <li>
        <p>
          Ordered list 2
        </p>
        <p>
          <a
            href="linktosomething.com"
          >
            Link to somethign
          </a>
        </p>
        <hr
          class="c4"
        />
        <p>
          <code
            class="c5"
          >
            inline code
          </code>
        </p>
        <pre
          class="c6"
        >
            const blockCodeFunc = () =&gt; {
    return 'works!';
  }
        </pre>
      </li>
    </ol>
  </div>
</div>
`;

exports[`Markdown Renders Markdown with Interactive React Components 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 {
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border-radius: 0px;
}

.c2 {
  height: 16px;
}

<div>
  <div
    class="c0 c1"
  >
    <div
      class="c0 c1"
    >
      <p>
        Markdown Text
      </p>
    </div>
    <div
      class="c2"
      height="16px"
    />
    <button
      data-testid="button"
    >
      Click Me
    </button>
  </div>
</div>
`;
